First preliminary version of Schemes.jl.

HechtiDerLachs opened the pull request:
This is more of a roadmap but a complete implementation, yet. However, feel free to play around with it. Comments and questions are welcome.

HechtiDerLachs committed:
First preliminary version of Schemes.jl.

thofma commented:
Tests are failing. There is already an `identity` function in julia Base defined for all objects, which clashes with your definition. Easiest solution is to pick another name.

HechtiDerLachs commented:
Oops.

Yes, I'll probably import the indentity from base and simply extend it for my new structs. 

Sorry! I didn't check for all other tests. Hope this didn't cause too much trouble now.

thofma commented:
I don't think we should overload `identity` to doing something else. It is quite useful to have that  `identity(x) === x` no matter the input (see https://docs.julialang.org/en/v1/base/base/#Base.identity). The name might be a bit unfortunate, but we have to live with it.

HechtiDerLachs committed:
Minor fixes due to name conflict for 'identity'

HechtiDerLachs committed:
Merge branch 'oscar-system:master' into Schemes_for_Oscar

HechtiDerLachs commented:
I just fixed the naming conflict. Since 'identity' really is the identity function itself in Oscar, I replaced it by 'get_identity'. 

Now, how do I update the pull request to also comprise these latest changes?

fingolfin commented:
@HechtiDerLachs if you push changes to your branch on your repository on GitHub, then this PR automatically is updated to show these.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+
+####################################################################
+#  
+#  Miscalleneous routines used in the above 
```
```suggestion
#  Miscellaneous routines used in the above 
```

HechtiDerLachs replied:
Haha! True. One of the last few difficult words for me...

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+  vars = vcat( old_vars, new_vars )
+  # place the new variables up front so that when they're used in 
+  # localizations, they're eliminated first.
+  #vars = vcat( new_vars, old_vars )
```
This is confusing: you explain something, but then the code doing it is commented out. Any of the following would be better:
- if you don't plan on using it, just remove this
- if you would like to use it, keep it, but change the comment to explain why this is not actually done; e.g. "TODO: change this so the new variables are up front that [...]. This is currently disabled because ... [ insert reason]"

HechtiDerLachs replied:
Yes, that's confusing. The comment was due to some intermediate decision that was then taken back.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+  # localizations, they're eliminated first.
+  #vars = vcat( new_vars, old_vars )
+  S, v = PolynomialRing( k, vars )
+  # @show S, v
```
Do you really need this debug code so often that it's worth leaving it in here?
```suggestion
```

HechtiDerLachs replied:
Amended.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+##################################################################
+#
+# Checks whether some power of u is contained in the principal ideal 
+# generated by g and returns the elements (k,a) of the equation 
```
```suggestion
# generated by g and returns a solution (k,a) of the equation 
```

HechtiDerLachs replied:
ok.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+# Checks whether some power of u is contained in the principal ideal 
+# generated by g and returns the elements (k,a) of the equation 
+#    u^k = a*g
+# If no such equation holds, it returns nothing.
```
```suggestion
# If no such solution exists, it returns `nothing`.
```

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+#    u^k = a*g
+# If no such equation holds, it returns nothing.
+function coeffs_in_radical( g::MPolyElem, u::MPolyElem )
+  k = Int(0);
```
```suggestion
  k = 0
```

HechtiDerLachs replied:
I don't remember this in detail right now, but I'm afraid for some reason, this was necessary. Either way: This routine is not yet implemented really. I got tired at the point where I didn't know how to do an infinite for-loop in Julia and I got bored looking that up. The whole implementation should be put under the TODO label. So, I'm not changing this for now, but certainly later.

thofma replied:
At the moment it will silently return wrong results?

fingolfin replied:
Yeah I don't understand this argument: if this code is severely broken, then remove it or comment it out or change the first line to `error("TODO this does not yet work")`? Or, just fix it along the lines I suggested?

HechtiDerLachs replied:
> At the moment it will silently return wrong results?

I'm afraid, yes. Is this unacceptable in the experimental section? Either way, it's intended to be an internal routine that is not being used, yet.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+  R = parent(g)
+  if parent(g) != parent(u) 
+    error( "elements are not contained in the same ring!" )
+  end
```
More Julia-ish would be to write this:
```suggestion
  parent(g) == parent(u) || error( "elements are not contained in the same ring!" )
```

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+  end
+  success=true
+  a=zero(R)
+  for k in (1:100) # Todo: replace by some infinite range eventually
```
```suggestion
  for k in 1:100 # Todo: replace by some infinite range eventually
```
What do you mean, "some infinite range"??? If you just want to do an infinite loop, you can do so right now, via `while true ... end`. So what's needed for this replacement?
```suggestion
  while true
```

HechtiDerLachs replied:
Is this really the way to go? Then I have to care about initialization and iteration of my counter manually?

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+  if !radical_membership( u, ideal( parent(g), g ) )
+    return nothing
+  end
+  success=true
```
```suggestion
```

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+  success=true
+  a=zero(R)
+  for k in (1:100) # Todo: replace by some infinite range eventually
+    (success, a) = divides( g, u )
```
```suggestion
    success, a = divides( g, u )
```

fingolfin replied:
You very often put unnecessary parenthesis around either ranges (i.e. `(1:n)` instead of just `1:n`) or multi value returns. While this works, it is unusual for Julia code.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+    end
+    u = u*u
+  end
+  return (2^k, a)
```
You don't know that k will be less than 63, do you? So better to be safe and return an fmpz
```suggestion
  return fmpz(2)^k, a
```

HechtiDerLachs replied:
Ok, I'll keep that in mind once we get there.

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+mutable struct OrderedMultiindex <: Multiindex 
+  n::Int
+  p::Int
+  index::Vector{Int}
```
What do these entries mean? Perhaps add at least some *minimal* comments?

HechtiDerLachs replied:
Fair. Documentation for this file is either way rather poor by now, I'm afraid...

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+      v[i] = i
+    end
+    #@show n, p, v
+    return new( n, p, v )
```
```suggestion
    return new( n, p, collect(1:p) )
```

HechtiDerLachs replied:
This, I don't get: What is the 'collect' doing? I intend to store a vector of integers and that's what v is. Doesn't 'collect' convert this to something like a set?

thofma replied:
No, `collect(v)` yields`[i for i in v]`, so for `1:n` the list `[1,...,n]`.

HechtiDerLachs replied:
But again: Why would I bother calling another routine here if the object I have is already of the form that I want it to be?

thofma replied:
Your four lines of code (line 18 to 21) are literally the same thing as `collect(1:p)`. So why not replace it with `collect(1:p)`?

fingolfin replied:
The advantage of `collect(1:p)` is that it is concise, making it easier to read your code

HechtiDerLachs replied:
Ah! Now I get it! I didn't understand that collect is supposed to replace the other lines of code, but I thought it was just a conversion. Nice, then let's do that.

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+  outstr= "0"
+  for k in 1:length(i.index) 
+    outstr = outstr * " < " * string(i.index[k]) 
+  end
```
```suggestion
  for k in i.index
    outstr *= " < $k"
  end
```

Alternatively you could use `join(i.index, " < ")`

HechtiDerLachs replied:
This operator exists in Julia? I thought, it didn't. If yes, I'll be happy to apply it.

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+
+function Base.iterate( i::OrderedMultiindex ) 
+  for k in 1:i.p
+    i.index[k] = k
```
Wait, this *mutates* the object over which you are iterating! That seems completely wrong?!?

HechtiDerLachs replied:
Really? From what I was reading about implementing the iteration interface, this was the way to go. Also, you don't iterate over a range of integers without changing the value of your variable, don't you?

fingolfin replied:
No, of course an iterator should not modify the object it "iterates over". That's what the "state" variable is for -- it stores the "state" of the iteration.

To take your example of "iterating over a range of integers": what you are doing is akin to *modifying the range of integers* over which you iterate.

thofma replied:
(It is not forbidden to have stateful iterators in julia.)

fingolfin replied:
True, but it wasn't clear to me that `OrderedMultiindex` is an iterator type? If so, then it should be renamed to make that clear.

HechtiDerLachs replied:
> No, of course an iterator should not modify the object it "iterates over". That's what the "state" variable is for -- it stores the "state" of the iteration.
> 
> To take your example of "iterating over a range of integers": what you are doing is akin to _modifying the range of integers_ over which you iterate.

I see your point. However, if you look at the design of the multiindex (which is admittedly very poorly commented), then you see that the range is fixed by the values of n and p. The i (or a) accounts for the state variable. And this is the only thing that is being touched. 

However, I will review this regarding the question whether the state variable can be used for this.

thofma replied:
I think it fine that this is the iterator type itself and storing the state inside can be done. But then, the first return value should not be the iterator itself, but maybe a `Vector` representing the ordered index itself? Because this is what the `a` is in `for a in i`.

HechtiDerLachs replied:
But inside the loop one will want to do something with the index a. And as I said before, the set of increasing integers in a are not worth much unless one also provides the context n and p. So I preferred to have the whole object returned. Which has the further advantage that one can use the routines developed for handling ordered Multiindices and one does not have to do things by hand on a vector of integers. 

But I feel this is for a longer discussion that we probably don't want to have here, typing everything.

thofma replied:
I see, sure! Then there should be two distinct types. The iterator objects as well as the things that are returned, which are the multiindices themselves. Mixing both will probably lead to problems down the road, since this is not how iterators are supposed to be implemented in julia.

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+  return i, i.index
+end
+
+function Base.iterate( i::OrderedMultiindex, state ) 
```
You are not using `state` in this function, which further indicates something is wrong with the iterator implementation.

HechtiDerLachs replied:
I think using 'state' is optional.

fingolfin replied:
No

thofma replied:
Is the `OrderedMultiindex` supposed to represent one index `a1 < ... < ak` or the set of all of these indices? I am confused because the object itself is returned from the iterator.

HechtiDerLachs replied:
Both (see above): It has the datum for the range stored in n and p and then the datum for the actual state in i (resp. a). 

Keeping the knowledge about n and p is important in order to make sense of the values for a and, for instance, to calculate the associated linear index, etc. I'll try to write some more explanatory comments in the source file and then hopefully things will be more clear.

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+  function HomogMultiindex( n::Int, d::Int, a::Vector{Int}, i::OrderedMultiindex )
+    if n <= 0 
+      error( "invalid number of variables" )
+    end
```
again, more Julia-like wold be something like this (I won't mark this everywhere :-) ); I also improved the error message to give a better hint what's wrong
```suggestion
    n > 0 || error("number of variables must be positive")
```

HechtiDerLachs replied:
Ok, that's shorter, for sure. I'll try to adapt that style in the future. And I think, I'll learn it, anyway.

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+  #@show alpha.i.index, alpha.i.p
+  if alpha.i.p > 0
+    a = [ alpha.i.index[1] - 1 ]
+    a = vcat( a, [ alpha.i.index[k+1] - alpha.i.index[k] - 1 for k in (1:alpha.i.p-1) ] )
```
This keeps copying data around. Better do this:
```suggestion
    append!( a, [ alpha.i.index[k+1] - alpha.i.index[k] - 1 for k in 1:alpha.i.p-1 ] )
```

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+  if alpha.i.p > 0
+    a = [ alpha.i.index[1] - 1 ]
+    a = vcat( a, [ alpha.i.index[k+1] - alpha.i.index[k] - 1 for k in (1:alpha.i.p-1) ] )
+    a = vcat( a, [ alpha.i.n - alpha.i.index[alpha.i.p] ])
```
see above

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+function Base.iterate( alpha::HomogMultiindex )
+  Base.iterate( alpha.i )
+  alpha.a = extract_exponent( alpha )
+  return alpha, alpha.a 
```
This iterator implementation also seems wrong: you are discarding the result of `Base.iterate`

HechtiDerLachs replied:
It's adding up...(haha, pun) No, but seriously: Maybe I have to review this. It's working, though. I need to find someone who has already implemented iterators to talk to about this.

fingolfin replied:
Feel free to discuss with me in gather.town or slack; others there will also be able to help

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+
+# Use Ordered multiindices for indexing 2-dimensional arrays
+
+function Base.getindex( v::Array{Any,2}, α::OrderedMultiindex )
```
```suggestion
function Base.getindex( v::Matrix{Any}, α::OrderedMultiindex )
```
We have shifted away from using `Array{T,2}` to its alias `Matrix` (which is how Julia prints the type starting with Julia 1.6).

Also, I am not sure if this signature is really what you want: this will accept only `Matrix{Any}` but not e.g. `Matrix{Int}`. To accept any `Matrix{T}`, do this:
```suggestion
function Base.getindex( v::Matrix, α::OrderedMultiindex )
```

HechtiDerLachs replied:
Ok, good to know. Thx.

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+# Use Ordered multiindices for indexing 2-dimensional arrays
+
+function Base.getindex( v::Array{Any,2}, α::OrderedMultiindex )
+  if α.p < 2 
```
I am not sure all of us are comfortable with using unicode in our source files (I don't mind much, but it can be hassle if one has no easy way to type these letters; one then has to essentially copy&paste these identifiers to edit the code, which is really annoying.

Perhaps you could just call it `a`?

simonbrandhorst replied:
Yes please, I do not want to spend my day googling how to type alpha (or to start learning unicodes by heart).

thofma replied:
Same with `Γ` below.

HechtiDerLachs replied:
But this is a very nice feature of Julia! 

It works like this: You type the latex command, say \Gamma, and then you hit the TAB-key. Simsalabim: There comes your unicode character! 

In this particular place, it can be avoided, but in general I think it can significantly increase readability since it's much closer to the mathematical way of writing things.

fingolfin replied:
But I don't use Julia to type in my code. And when I need to edit this Julia file, I'll now have to deal with those special letters, and in most editors, they are *not* convenient to enter.

HechtiDerLachs replied:
Then, I think, one should use a proper editor... like Vim with the Julia plugin.

No, jokes aside. I see that there are reasons to keep the actual code in ASCII letters. But for explanatory comments I will keep using the unicode feature. In the end, the use of such letters is intended by design of the language and in my opinion, people should not blame this on their editors, but adjust their working environment to the language that they would like to work with.

thofma replied:
We are trying to find/found a minimal common denominator to work with in this project and this is not helpful in this regard.

HechtiDerLachs replied:
Ok, I see. Even in the comments? I saw this in the code of the intersection theory module and I was delighted how nice everything looked.

thofma replied:
I don't mind it in the comments. But be prepared that it might go away once someone has to update the comments to reflect changes in the code. Not sure what @fingolfin thinks about unicode in comments.

fingolfin commented on experimental/Geometry/Schemes.jl:
```diff
@@ -0,0 +1,1244 @@
+import AbstractAlgebra: FPModule, FPModuleElem
+
+include( "./Multiindices.jl" )
+using Oscar.Multiindices
```
If you end up doing `using FOO` immediately anyway, why even have `module FOO` in the first place?

HechtiDerLachs replied:
I don't really understand...

At this point I probably just fiddled around with the different methods of inclusion just to get things to run. I think, if I did not include the file explicitly, then Oscar just wouldn't know about this module and it couldn't be used. But someone should teach me how to do these things properly at some point.

fingolfin commented on experimental/Geometry/Schemes.jl:
```diff
@@ -0,0 +1,1244 @@
+####################################################################################
+# The classical affine scheme, explicitly given as the quotient 
+# ring of a polynomial algebra.
+mutable struct Spec{S,T,U} <: AffineScheme{S,T,U}
```
You can restrict the acceptable type parameters here
```suggestion
mutable struct Spec{S<:Ring,T<:MPolyRing,U<:MPolyElem} <: AffineScheme{S,T,U}
```

HechtiDerLachs replied:
Ok... But have they not already been restricted in the declaration of AffineScheme? Why do that twice?

fingolfin replied:
Sorry, I missed the definition of `AffineScheme` somehow.

fingolfin commented on experimental/Geometry/Schemes.jl:
```diff
@@ -0,0 +1,1244 @@
+    if k != coefficient_ring(R)
+      error( "Base ring of the affine scheme does not coincide with the base ring of the associated algebra." )
+    end
+    # TODO: Implement further plausibility checks to be performed at runtime.
```
Indeed, at the very least, check that the concrete types `U` and `T` "fit"

fingolfin commented on experimental/Geometry/Schemes.jl:
```diff
@@ -0,0 +1,1244 @@
+  function Spec(k::S, R::T, I::MPolyIdeal{U} ) where{
+			S <: Ring, T <:MPolyRing , U <: MPolyElem}
+    if k != coefficient_ring(R)
+      error( "Base ring of the affine scheme does not coincide with the base ring of the associated algebra." )
```
No trailing periods in error message
```suggestion
      error("Base ring of the affine scheme does not coincide with the base ring of the associated algebra")
```

HechtiDerLachs replied:
Ok. Again, good to know.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+module Misc
+
+using Oscar
```
This shouldn't be necessary?!

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,56 @@
+####################################################################
+
+
+function add_variables( R::MPolyRing, new_vars::Vector{String} )
```
This one is very nitpicky, but: for future code, please don't use whitespace right of `(` and left of `)`. I.e.:
```suggestion
function add_variables(R::MPolyRing, new_vars::Vector{String})
```

Anyway, this is a minor thing. Perhaps we should just setup a default format setting for `JuliaFormatter`, and then ask everyone to use that for formatting their code... but it's really not important. So don't worry about this :-)

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+  if n != alpha.n
+    error( "number of variables not compatible." )
+  end
+  result = 1
```
This code means that `result` in general changes its type: it always starts out as an `Int`, but then it is (probably) changed to some object of the same type as the members of `x`...

Better would be to set it to `one(x[1])`, I think. Or use `prod` (see another code change suggestion by me)

HechtiDerLachs replied:
I see...

So writing a routine in Julia for a non-specified type is OK, but altering the type of some variable within the body of the routine is not OK? 

Yes, one problem is that it returns Int(1) in case of exponent zero which might cause the calling routine to fail.

fingolfin commented on experimental/Geometry/Multiindices.jl:
```diff
@@ -0,0 +1,221 @@
+  for k in (1:n)
+    result = result * x[k]^(alpha.a[k])
+  end
+  return result
```
```suggestion
  return prod(k -> x[k]^alpha.a[k], 1:n)
```

HechtiDerLachs replied:
Huh! Candy!

fingolfin commented on experimental/Geometry/Schemes.jl:
```suggestion
    x = new{S,Td,Ud,Tc,Uc}(domain, codomain, pullback)
```

HechtiDerLachs replied:
That's also how I did this originally. The format used here I learned from Simon. Is there a performance reason to favor the one-liner?

fingolfin replied:
It's more idiomatic, and more concise.

fingolfin commented on experimental/Geometry/Schemes.jl:
```suggestion
  if isdefined(f, :imgs_frac))
```

HechtiDerLachs replied:
I already went through the code to catch these. But some of them seem to have slipped my notice...

fingolfin commented on experimental/Geometry/Schemes.jl:
```suggestion
  append!( fracs, [ 1//g for g in den ])
```
And similar elsewhere

fingolfin commented on experimental/Geometry/Schemes.jl:
```suggestion
  Base.println( io, "Glueing of" )
  Base.println( io, Γ.firstPatch )
  Base.println( io, "and" )
  Base.println( io, Γ.secondPatch )
  Base.println( "along" )
```

thofma replied:
Also missing an `io`.

fingolfin commented on experimental/Geometry/Schemes.jl:
```suggestion
  glueings::Matrix{Union{Glueing,Nothing}}     # A list of glueings between the patches 
```
... and similar in many other places

fingolfin reviewed (commented):
Thank you for working on this!

Now, fair warning: I am going to leave a lot of comments on this PR; many of them are nitpicky. Please do not feel "put down" by these comments -- the reason I am writing them (and investing considerable time in doing so) is that I am hoping I can convey some helpful things about Julia and OSCAR, which will save time and work down the road. Also, please feel free to question any of my remarks, or ask for further explanations. And for some of my comments, people like @fieker or @thofma may disagree, so feel free to get a "3rd opinion" from them for any of this :-).

fieker commented:
On Tue, Aug 31, 2021 at 01:59:40AM -0700, simonbrandhorst wrote:
> @simonbrandhorst commented on this pull request.
> 
> 
> 
> > +  end
> +  result = 1
> +  for k in (1:n)
> +    result = result * x[k]^(alpha.a[k])
> +  end
> +  return result
> +end
> +
> +function linear_index( alpha::HomogMultiindex ) 
> +  return linear_index( alpha.i )
> +end
> +
> +# Use Ordered multiindices for indexing 2-dimensional arrays
> +
> +function Base.getindex( v::Array{Any,2}, α::OrderedMultiindex )
> +  if α.p < 2 
> 
> Yes please, I do not want to spend my day googling how to type alpha (or to start learning unicodes by heart). 
> 
On my windows laptop, half of the alphas are not readable at all - they
look like a random square...

Fun apart: I though no-one was evern going to do
[alpha^k for k = 1..n]
This turns an algorithm for n multiplication potentially into one for
n^2/2 multiplications...
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/638#discussion_r699126052

simonbrandhorst commented:
The tests should be moved correspondingly to `test/experimental/Geometry/Schemes.jl `

fingolfin commented:
To be clear: I also have more comments, bit I thought it would be best to address those I pointed out (and of course in spots, not just those I marked), then have a fresh look at the adjusted code...

HechtiDerLachs committed:
Update experimental/Geometry/Misc.jl

Co-authored-by: Max Horn <max@quendi.de>

HechtiDerLachs committed:
Minor changes suggested by Max Horn.

HechtiDerLachs committed:
Update experimental/Geometry/Misc.jl

Co-authored-by: Max Horn <max@quendi.de>

HechtiDerLachs committed:
Update experimental/Geometry/Multiindices.jl

Co-authored-by: Max Horn <max@quendi.de>

HechtiDerLachs committed:
Turns out there was a problem with the last suggestion in the end.

HechtiDerLachs committed:
Update experimental/Geometry/Multiindices.jl

Co-authored-by: Max Horn <max@quendi.de>

HechtiDerLachs committed:
Update experimental/Geometry/Multiindices.jl

Co-authored-by: Max Horn <max@quendi.de>

HechtiDerLachs committed:
Minor changes.

HechtiDerLachs committed:
Merge branch 'Schemes_for_Oscar' of github.com:HechtiDerLachs/Oscar.jl into Schemes_for_Oscar

HechtiDerLachs committed:
Update experimental/Geometry/Multiindices.jl

Co-authored-by: Max Horn <max@quendi.de>

HechtiDerLachs committed:
Update experimental/Geometry/Multiindices.jl

Co-authored-by: Max Horn <max@quendi.de>

HechtiDerLachs committed:
Update experimental/Geometry/Misc.jl

Co-authored-by: Max Horn <max@quendi.de>

fingolfin commented:
@HechtiDerLachs i am not sure if adding my remarks one by one is the best way to go. First off, this would normally create a HUGE strain on our continuous integration, as each of your commits would trigger a full separate CI run. Better to "add to queue" those suggestions, and commit multiple at once.

Luckily in this case, since this is your first commit, you still need to be approved for running CI tests, so I'll wait with doing that until you are done...

But also note that many of my suggestion point out an issue in 1-2 spots, but really should be addressed in many more. It might be more efficient to locally work on these files and use an editor with "search and replace" to address many of these comments in bulk, and in all relevant places?

HechtiDerLachs committed:
Update experimental/Geometry/Schemes.jl

Co-authored-by: Max Horn <max@quendi.de>

HechtiDerLachs commented:
> @HechtiDerLachs i am not sure if adding my remarks one by one is the best way to go. First off, this would normally create a HUGE strain on our continuous integration, as each of your commits would trigger a full separate CI run. Better to "add to queue" those suggestions, and commit multiple at once.
> 
> Luckily in this case, since this is your first commit, you still need to be approved for running CI tests, so I'll wait with doing that until you are done...
> 
> But also note that many of my suggestion point out an issue in 1-2 spots, but really should be addressed in many more. It might be more efficient to locally work on these files and use an editor with "search and replace" to address many of these comments in bulk, and in all relevant places?

Alright. I'll look into that how to gather the comments in one thing then. Working on the necessary changes in my local repo in a general sense will then happen afterwards. For now, it was just too tempting to simply hit the 'apply suggested changes' button.

HechtiDerLachs commented:
> On my windows laptop, half of the alphas are not readable at all - they look like a random square... Fun apart: I though no-one was evern going to do [alpha^k for k = 1..n] This turns an algorithm for n multiplication potentially into one for n^2/2 multiplications...

I don't get this. Where did you find this in the code? If it's there, of course, it should be changed! But I don't recall having programmed anything like that.

fingolfin commented on experimental/Geometry/Schemes.jl:
```diff
@@ -0,0 +1,1244 @@
+###############################################################################
+# Getter functions
+
+function PrincipalSubScheme(parent, denom)
```
Only types (and by necessity, constructors) should use `CamelCase`; functions should use `snake_case`.

Since this distinction may be unclear and fuzzy for end users, and also sometimes developers, we also tend to have alternate constructors using `snake_case`; I can't judge whether this would make sense here (i.e. having a `spec_principal_open` function) but what I can say is that `PrincipalSubScheme` isn't a type, so it shouldn't have this name.

HechtiDerLachs replied:
> The tests should be moved correspondingly to `test/experimental/Geometry/Schemes.jl `

I don't find any such folder: 'test/experimental'. Are you sure?

simonbrandhorst replied:
As far as I know the directory structure of the test files should mirror that of the source code, so that we know where to look for the tests. If that folder is not there, you can create it.

fingolfin replied:
We haven't been very consistent about this, to be fair. E.g. the plane curves code by @delphinepol is now in `experimental` but the tests are still in `test/Examples/PlaneCurve-test.jl`.
 
So I think this is not *that* important right now...

But what I *do* wonder is whether perhaps `Geometry` should be changed to `AlgebraicGeometry`? We already have "Polyhedral Geometry" (and who knows, one day we might also have Differential/Incidence/Arithmetic/... geometry?).

HechtiDerLachs replied:
My point about the folder was that not only "Geometry" doesn't exist there, but really the whole test folder for the experimental section can't be found. It seems that nobody who ever contributed to the experimental section put their tests anywhere else but in the usual test folder.

HechtiDerLachs replied:
About the name: I have no strong feelings about the change there. AlgebraicGeometry would be fine with me.

fingolfin commented on experimental/Geometry/Schemes.jl:
```diff
@@ -0,0 +1,1244 @@
+end
+
+function parent(S::AffineScheme)
+  return S
```
I don't think this is compatible with the meaning of `parent` we employ elsewhere: elements have parents. But e.g. a submodule doesn't have a parent, it has a `supermodule`.

That said, in GAP, we'd also use `parent` in the same way you use it here. I could live with that, but I am not sure if there are strings attached...

So perhaps this should be `superscheme`? Or perhaps we can relax and allow `parent`? Perhaps @fieker @thofma @wbhart have thoughts on this?

HechtiDerLachs replied:
Good point. And close to the 'identity'-problem encountered earlier. We'll come up with something there. But yes: Suggestions are very welcome.

HechtiDerLachs committed:
Resolve one conflict and remove trailing characters form error messages.

fingolfin commented on experimental/Geometry/Schemes.jl:
These names (`get_foo`) are not idiomatic (e.g. we don't have `get_parent`, but rather `parent`).

simonbrandhorst replied:
The drawback of the more idiomatic approach is that one has to remember the order of the arguments. But either way I do not really care.

fingolfin replied:
I don't understand, there's only one argument here? And how does `get_foo` vs. `foo` help to indicate the order of arguments?

simonbrandhorst replied:
It was about the 
""" experimental/Geometry/Schemes.jl
    x = new{S,Td,Ud,Tc,Uc}()
    x.domain = domain
    x.codomain = codomain
    x.pullback = pullback
@fingolfin fingolfin 1 hour ago Member

It's more idiomatic, and more concise.""" comment. There was no answer button there. Sorry.

fingolfin commented:
Ooops, a file `experimental/Geometry/.Schemes.jl.swp` crept in.

I recommend to add `*.swp` (and `*~` and a bunch more) to your global gitignore list (see e.g. https://gist.github.com/subfuzion/db7f57fff2fb6998a16c or https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/  or https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer)

HechtiDerLachs commented:
Oh damn! No idea how that happened. Probably while adding a whole folder at once. I'll take care of that.

HechtiDerLachs committed:
Remove the .swp file from the repository.

HechtiDerLachs committed:
Implement the divides_power routine (former radical membership test with coefficients).

HechtiDerLachs committed:
Amend suggested changes by Max Horn.

HechtiDerLachs committed:
Amend changes suggested by Max Horn.

HechtiDerLachs committed:
Add a radical membership test.

HechtiDerLachs committed:
Remove the non-idiomatic 'get_' from the functions.

HechtiDerLachs committed:
Extend radical membership to quotient rings.

HechtiDerLachs committed:
Replace the greek letters within the source code.

HechtiDerLachs committed:
Fix a twist in Rabinowitsch's trick

HechtiDerLachs committed:
Intermediate commit while working on the conversion of the fractional representation. There seems to be a bug for 'divides' in quotient rings (Issue #645) that needs to be addressed first.

HechtiDerLachs committed:
Merge branch 'oscar-system:master' into Schemes_for_Oscar

HechtiDerLachs committed:
Merge branch 'Schemes_for_Oscar' of github.com:HechtiDerLachs/Oscar.jl into Schemes_for_Oscar

HechtiDerLachs committed:
Finish first prototype for conversion of fractional representation to algebra homomorphism for morphisms of affine schemes.

HechtiDerLachs committed:
Make the localization at units not introduce superfluous variables.

HechtiDerLachs committed:
Remove some debugging lines and streamline the inclusion methods for SpecPrincipalOpen.

HechtiDerLachs committed:
Start working on sections of vector bundles.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,202 @@
+  if length(G)==1 && G[1] == one(S)
+    return true
+  end
+  return false
```
Can't this be simplified to this? (Or do we loose performance somehow? But if so, shouldn't we improve the `ideal_membership` method?)
```suggestion
  return one(S) in I
```

HechtiDerLachs replied:
Good point. I think in principal we should rely on the ideal membership being implemented in more efficient ways, than oneself could possibly do it on the fly. So the suggested change seems correct. 

However, I would like to first do some performance tests with different implementations of this routine: Groebner-based vs. divides-based, etc. and then see what we really want to use. So we might switch to using the coefficients of the Groebner-basis methods in the end...

Btw: Did anyone yet think of providing ideal membership together with the coefficients writing the element as linear combination of the chosen generators? That would be nice to have, but again tend to contradict the general Oscar approach to hide away the algorithmic backside (as also discussed on Slack).

hannes14 replied:
Representing an element in terms of the generators canbe done by liftstd (from Singular.jl),
but is much more expensive than just a membership test:
Membership test can stop as soon as the ledading term cannot be reduced further,
liftstd requires a full reduction AND keeping track of all operations

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,202 @@
+  if length(G)==1 && G[1] == one(S)
+    return true
+  end
+  return false
```
```suggestion
  return one(S) in I
```

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,202 @@
+  R = parent(f)
+  if R != parent(g) 
+    error( "Polynomials do not belong to the same ring" )
+  end
```
```suggestion
  R == parent(g) || error("Polynomials do not belong to the same ring")
```
... and similar in many other places

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,202 @@
+# Test whether some power of the polynomial f is contained in the 
+# principal ideal generated by g
+#
+function Oscar.radical_membership( f::U, g::U ) where{ U<:MPolyElem }
```
```suggestion
function radical_membership(f::U, g::U) where {U<:MPolyElem}
```
and similar elsewhere

fingolfin replied:
Also, just wondering why the existing method `radical_membership(f::T, I::MPolyIdeal) where T <: MPolyElem` can't be used? I.e.
```suggestion
radical_membership(f::U, g::U) where {U<:MPolyElem} = radical_membership(f, ideal([g]))
```

HechtiDerLachs replied:
Yes, that might work fine. But the way we want to go here is still kind of open and at some point I probably wanted a little more access to what's happening in the background. We'll see about this once we do the review.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,202 @@
+  # Test for radical membership in the first place.
+  # TODO: Doesn't this computation already allow us to read 
+  # off our desired coefficients? How expensive is this compared 
+  # to what follows?
```
It does indeed allow you read it off (this is e.g. described in "Ideals, varieties, and algorithms" by Cox,Little and O'Shea). Of course to do that, we can't do `one(R) in I` as I suggested above, you need the actual coefficients in the GB...

HechtiDerLachs replied:
exactly. And the question is what is the "Oscar"-way of doing that. See the discussion on Slack.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,202 @@
+  k = base_ring(R)
+  old_vars = String.( symbols(R) )
+  n = length( old_vars )
+  vars = vcat( old_vars, new_vars )
```
Elsewhere we accept both strings and symbols for variable names, so perhaps also here (and same for the other `add_variables` method below)?

```suggestion
add_variables(R::T, new_vars::Vector{String}) where {T<:MPolyRing} = add_variables(R, Symbol.(v))

function add_variables(R::T, new_vars::Vector{Symbol}) where {T<:MPolyRing}
  k = base_ring(R)
  old_vars = symbols(R)
  n = length( old_vars )
  vars = vcat( old_vars, new_vars )
```

HechtiDerLachs replied:
Ok, that makes sense. I'll adjust that.

fingolfin commented on experimental/Geometry/Misc.jl:
```diff
@@ -0,0 +1,202 @@
+  I = ideal(S, 
+	    vcat( [phi(lift(g)), one(S)-t[1]*phi(lift(f))],
+		 [phi(h) for h in gens(modulus(P))] )
+	    )
```
Aren't you basically lifting the quotient ideal defining P to R and adding g in, and then performing a `radical_membership` test on that? Of course then it's not for a principal ideal. But we already have a `radical_membership` method for general ideals... Can't we just call that?

HechtiDerLachs replied:
Probably yes.

HechtiDerLachs committed:
Fix a bug on the ordering of denominators and units for localized affine schemes.

HechtiDerLachs committed:
Commit zur Besprechung.

HechtiDerLachs committed:
Merge branch 'oscar-system:master' into Schemes_for_Oscar

HechtiDerLachs committed:
Minor changes after talk in the meeting.

HechtiDerLachs committed:
Add various copy constructors.

HechtiDerLachs committed:
Add some functionality for intersections.

HechtiDerLachs committed:
First part of rewriting using a single localization variable.

HechtiDerLachs committed:
Commit for exposing a bug that I couldn't reproduce otherwise.

thofma committed:
Some asserts

HechtiDerLachs committed:
Merge pull request #9 from oscar-system/reporting_a_possible_bug

Some asserts

HechtiDerLachs committed:
Fix the 'bug'.

HechtiDerLachs committed:
Fix the division issues.

HechtiDerLachs committed:
Merge branch 'oscar-system:master' into Schemes_for_Oscar

HechtiDerLachs committed:
Add docstrings.

HechtiDerLachs committed:
Move everything to a folder of its own.

HechtiDerLachs committed:
Adjust the main Oscar.jl file.

HechtiDerLachs committed:
First attempt on direct sums of vectorbundles.

thofma commented:
Is `diagonal_matrix` not doing what you want?

```julia
julia> diagonal_matrix([matrix(ZZ, 2, 2, [1, 2, 3, 4]), matrix(ZZ, 1, 2, [1, 2])])
[1   2   0   0]
[3   4   0   0]
[0   0   1   2]
```

HechtiDerLachs commented:
I recall that it somehow didn't. But now that I tried again some example in the repl, it seems to work. That would definitely be the first choice to use here, since the typing of the other methods was horrifying. 

So I'm probably going to change that again.

HechtiDerLachs committed:
Use the native matrix routines instead of our own diag.

HechtiDerLachs committed:
First attempt on documentation.

HechtiDerLachs committed:
Add an optional argument for naming the extra variable in Rabinowitsch's trick.

HechtiDerLachs committed:
Fundamentals for cartesian products of affine schemes.

lkastner commented on docs/src/Experimental/schemes.md:
````diff
@@ -0,0 +1,183 @@
+```
+
+Another important instance of affine schemes is 
+```@docs
+    SpecPrincipalOpen{S,T,U}
+```
+```@docs
+   parent(::SpecPrincipalOpen)
+```
````
You can just make one big multiline docs block, i.e.
````suggestion
```@docs
    SpecPrincipalOpen{S,T,U}
   parent(::SpecPrincipalOpen)
```
````
i.e. just join by leaving out the separating part.

fingolfin replied:
Indeed, that would be preferable

fingolfin commented on docs/src/Experimental/schemes.md:
````diff
@@ -0,0 +1,183 @@
+    CoveredScheme
+```
+
+A covering itself relies on 
````
I don't think this "filler texts" work that well in the final rendered version of the docs. So I'd just drop them, and perhaps instead insert one brief sentence or paragraph at the start of the section that mentions how these types fit together.

fingolfin commented on docs/src/Experimental/schemes.md:
````diff
@@ -0,0 +1,183 @@
+functions for the specified covering. 
+```@docs
+   covering(::LineBundle)
+```
+```@docs
````
```suggestion
```
... and similar in many other places

fingolfin commented on docs/src/Experimental/schemes.md:
````diff
@@ -0,0 +1,183 @@
+need to even cover an affine scheme by finer patches 
+in order to represent it as a line bundle of this form.
+
+With basically the same implementation one has 
+```@docs
+    VectorBundle
+```
+which we separated from the line bundles because of the 
````
Again, I don't think this works well in the rendered documentation at all: A sentence begins, then a big blob of other text is inserted, and then the sentence resumes -- that's usually annoying to read.

fingolfin commented on docs/src/Experimental/schemes.md:
````diff
@@ -0,0 +1,183 @@
+   transitions(::LineBundle)
+```
+Note that not every line bundle on an affine 
+scheme is necessarily trivial! In general one will 
````
```suggestion
scheme is necessarily trivial. In general one will 
```

fingolfin commented on experimental/Schemes/Misc.jl:
```diff
@@ -0,0 +1,249 @@
+####################################################################
+
+
+function add_variables( R::T, new_vars::Vector{String} ) where{ T<:MPolyRing } 
```
Please add at least a one line comment to your functions that explain what they do, even if they are internal and you don't want to add a docstring (though a proper docstring usually would be best, even for internal functions -- but I totally get not wanting to write one before it is clear that the function will stay).

I know the existing Oscar code is often bad in this place, but I'd like us to improve upon this standard.

fingolfin commented on experimental/Schemes/Misc.jl:
```diff
@@ -0,0 +1,249 @@
+  return (k, a)
+end
+
+# TODO: Up to now, this is merely a copy of the above routine. 
```
But what is the task to do?

fingolfin commented on experimental/Schemes/Misc.jl:
```diff
@@ -0,0 +1,249 @@
+@Markdown.doc """
+    function diag( d::Vector{MatSpaceElem} )
+
+Writes the matrices in the list d as blocks on the 'diagonal' of 
+a new matrix D and returns D. 
```
```suggestion
Writes the matrices in the list `d` as blocks on the 'diagonal' of 
a new matrix `D` and returns `D`. 
```
Why is "diagonal" quoted?

fingolfin commented on experimental/Schemes/Multiindices.jl:
```diff
@@ -0,0 +1,213 @@
+  Base.print( io, outstr )
+end
+
+function Base.iterate( i::OrderedMultiindex ) 
+  for k in 1:i.p
+    i.index[k] = k
+  end
+  return i, i.index
+end
```
These multi index types still suffer from the issues we discussed last time around, so until this is addressed, it seems pointless to try to review this PR again.

If you feel need to discuss this further, we can of course do so, but perhaps better on gather or in person.

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+polynomial algebra of type T over a base ring k of type S and I ⊂ R a finitely generated ideal 
+with elements of type U.
+"""
+mutable struct Spec{S,T,U} <: AffineScheme{S,T,U}
```
Why is this `mutable`?

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+
+The type parameter S is the type of the base ring k over which the Scheme is defined.
+"""
+abstract type Scheme{ S <: Ring }end
```
```suggestion
abstract type Scheme{S <: Ring} end
```

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+@Markdown.doc """
+    abstract type Scheme{ S <: Ring }
+
+The type parameter S is the type of the base ring k over which the Scheme is defined.
```
I think we agreed overall that argument names are in backticks, as it is code; that's also what Julia does in general.
```suggestion
The type parameter `S` is the type of the base ring `k` over which the scheme is defined.
```

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+  if base_ring(J) != X.R 
+    error( "Ideal does not belong to the ambient ring of the variety" )
+  end
+  Y = Spec( X.k, X.R, X.I )
+  Y.I = Y.I + J
+  return Y
```
```suggestion
  return Spec( X.k, X.R, X.I + J )
```

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+end
+
+# The copy constructors.
+# In this case, it's easy, since non of the used data structures 
```
```suggestion
# In this case, it's easy, since none of the used data structures 
```

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+# Simply storing v is wrong, because it is an object that could potentially
+# also be accessed from other parts of the program. If v is altered, then 
+# also the member variable of our Spec is altered! Thus in this case, 
+# we would really need to copy v before storing it in the Spec.
```
Indeed, that's standard in most high level languages, no? (Certainly in Python, JavaScript, Java, GAP, ...). Do we really need to discuss it at length here? I mean, it probably doesn't hurt, but it seems a bit peculiar?

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+#
+# Such behaviour is not expected for instances of rings or ideals. 
+# As far as I understand, these objects are supposed to behave like 
+# static ones to the user. But we should watch out for exceptions and 
```
What do you mean with "static" here? Immutable perhaps?

(Nb, there is a slight tension between "immutable" in the Julia sense, which I'd call "read-only", which is not recursive; and "immutable" in a more mathematical sense (that's how GAP uses it) which is kinda recursive and means "this will be the same mathematical object, nothing is gonna change that, although the object itself may change (i.e. may not be read-only) to store extra data (such as computed properties, additional charts, whatever)

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+
+@Markdown.doc """
+     defining_ideal(A::Spec)
+Returns the "defining ideal" I of the affine scheme A, where A = Spec R/I.
```
```suggestion

Returns the "defining ideal" I of the affine scheme A, where A = Spec R/I.
```

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+end
+
+@Markdown.doc """
+    function parent(S::AffineScheme)
```
```suggestion
   parent(S::AffineScheme)
```
The `function` must also be removed in a bunch of other docstrings below.

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+@Markdown.doc """
+    function is_parent_of( X::AffineScheme, U::SpecPrincipalOpen )
+
+Returns true whenever X occurs as a parent of U in the tree structure 
+for Spec and SpecPrincipalOpen.
```
That doesn't match what the code does; the comment before the docstrings seems to fit, though

fingolfin commented on experimental/Schemes/Schemes.jl:
```diff
@@ -0,0 +1,2587 @@
+
+Return the affine scheme corresponding to the ring $R/(0)$.
+"""
+function Spec( R::T ) where{T <: MPolyRing}
```
When you don't use the type `T` in the body, and also don't need to assert a relation between the types of different arguments, you can simplify the code:
```suggestion
function Spec(R::MPolyRing)
```

fingolfin commented on experimental/Schemes/Schemes.jl:
Err, what? That's (a) completely unclear (in how far does OSCAR's `compose` not follow a mathematical convention, and which is that???), and secondly sounds like a big no-no (namely, as if you are deliberately violating the specification of `compose`, which would be really bad, no matter that some code already does this, see issue #690)

fingolfin commented on experimental/Schemes/Schemes.jl:
```suggestion

@Markdown.doc """
```

fingolfin commented on experimental/Schemes/Schemes.jl:
```suggestion
    second_patch( G::Glueing )
```
No, the body of a one-line function still does not belong into the doc comment, please also fix this in other places

fingolfin commented on experimental/Schemes/Schemes.jl:
```suggestion
Note that one naturally has two realizations of this overlap as 
```

fingolfin commented on experimental/Schemes/Schemes.jl:
Why are you "afraid"? It sounds as if this was something nasty, undesirable -- but isn't it rather completely expected? I only mention this because I feel there is a deeper concern you have in mind here?!

fingolfin commented on experimental/Schemes/Schemes.jl:
```suggestion
    return C.patches[i]
```

fingolfin commented on experimental/Schemes/Schemes.jl:
Why not an out of bounds error? With the current code, `patch(C, -2)` or `patch(C, 2^60)` will work whenever there are no patches, but will fail once there are actually patches. That seems weird?

For that matter, why have this at all, when the user can also write `patches(C)[i]`?

fingolfin commented on experimental/Schemes/Schemes.jl:
Maybe also say what/where it is used internally? Presumably for constructing new coverings, yes? I am guessing you create a new covering object, then apply `_add_patch!` to it a couple times and return the result, after which it is not supposed to be modified like this again?

fingolfin commented on experimental/Schemes/Schemes.jl:
Ironic that you are not using `patch(C, i)` here ;-)

fingolfin commented on experimental/Schemes/Schemes.jl:
And yet another definition of `projective_space` :-). We absolutely need to sit down and discuss how to handle it.

fingolfin commented on experimental/Schemes/Schemes.jl:
```suggestion
  for j in 1:n
```

fingolfin commented on experimental/Schemes/Schemes.jl:
Broken indentation (or perhaps tabs mixed with spaces? amounts to the same to me, though...)


Perhaps most efficient would be if you run something like `JuliaFormatter` once on this file

fingolfin reviewed (commented):
I had another look at this, as last Friday it sounded as if this PR is now ready for review again and was indeed waiting to be reviewed and merged -- something that was not clear to me (we need to find a way to communicate this clearer).

I can't really comment much on the mathematics, though, it'd be good if e.g. @wdecker had a look at that.

But there are a bunch of technical remarks I can make, which would imply a ton of (mostly minor) changes;  I also certainly didn't look at all of the code -- this PR is too big by now for a proper review, I am afraid. Perhaps we can come up with ways to break it into chunks that can be reviewed and merged separately?

HechtiDerLachs commented:
Thanks for all the new input! 

Indeed, I could try to break this into smaller chunks. Either way, we were thinking of splitting this into multiple files, since it has grown too big by now, already. 

Concerning the mathematics: There is not too much mathematics in there, yet. It's still more of a chassis with the engine missing. But that's about to change soon, hopefully.
